[gcal] Add io and persistence feature to OH2

mhilbush opened the pull request:
Signed-off-by: Mark Hilbush <mark@hilbush.com>

watou commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -656,6 +656,12 @@
 
+  <feature name="openhab-persistence-gcal" description="GCal Persistence" version="${project.version}">
+    <feature>openhab-runtime-base</feature>
+    <feature>openhab-runtime-compat1x</feature>
```
openhab-persistence-gcal depends on openhab-io-gcal1 (it imports the `org.openhab.io.gcal.auth` package)

mhilbush replied:
Ok. Dependency added.

watou commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -768,6 +774,12 @@
+    <feature>openhab-runtime-base</feature>
+    <feature>openhab-runtime-compat1x</feature>
+    <bundle start-level="80">mvn:org.openhab.io/org.openhab.io.gcal/${project.version}</bundle>
+  </feature>
```
Could you add a default `gcal.cfg` file similar to what is done on line 667?

watou replied:
The binding itself actually doesn't know where its configuration comes from.  The 1.x runtime used openhab.cfg, but it's refactored in OH2 to have a .cfg per service, gcal.cfg in this case, for text file configuration.  Or via Paper UI so no config file has to be used.  

To add gcal.cfg, just take the GCal section from openhab_default.cfg into features/openhab-addons-external/src/main/resources/conf/gcal.cfg (but remove the `gcal:` prefixes in it).  Modify `features/openhab-addons-external/pom.xml` like in #4814.

To allow GCal to be configured from the Paper UI, add bundles/io/org.openhab.io.gcal/ESH-INF/config/config.xml similar to #4049, and see [this documentation](https://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html#configuration-descriptions).

mhilbush commented:
I made the changes for gcal.cfg.  openhab_default.cfg contained the old (pre-OAuth) config parameters, so I just added client_id, client_secret, and calendar_name to gcal.cfg.

I didn't do the config.xml change, as I'd prefer to get this part right first.  :-)

watou commented on features/openhab-addons-external/src/main/resources/conf/gcal.cfg:
```diff
@@ -0,0 +1,10 @@
+#client_secret=
+
+# The name of the Google Calendar that GCal should use
+#calendar_name=
```
Looks good, but I worry that users might not know which are required or optional.  So a bit more information as described [here](https://github.com/openhab/openhab/wiki/GCal-Binding#obtain-the-credentials) might avoid head scratching or hair pulling.  Also, when you've updated the helpfulness of the comments a bit, could you also fix openhab_default.cfg to match (adding back the `gcal:` prefixes in that version)?

mhilbush replied:
Sure. Good point. I didn't want to repeat what was in the Wiki, but I certainly can add more to the comments.

As for openhab_default.cfg, I'm not sure what you're asking as I didn't change that file.  It still contains the gcal: prefixes.

watou replied:
> openhab_default.cfg contained the old (pre-OAuth) config parameters

I was hoping this PR would be a good opportunity for consistency between the now two .cfg files, seeing as you report that we missed updating openhab_default.cfg file with @SirAd's big refresh.

mhilbush replied:
Ah, I see now.  :-/  

I wasn't sure if that file was supposed to be updated.  I can do that.  Note that there also are some gcal-persistence parameters in openhab_default.cfg.  Can those be added to gcal.cfg (in a separate section of the file explaining their use for the persistence feature), or is a separate gcal-persistence.cfg needed?
```
#gcal-persistence:offset=
#gcal-persistence:executescript=
```
What about the Wiki?  It still references openhab.cfg and gcal: prefixes?

And, I see there are some log messages in the code that reference openhab.cfg...

mhilbush replied:
Actually, looking at this more closely, GCal Persistence has it's own calendar_name parameter, which may dictate a separate cfg file to avoid conflicting with GCal's calendar_name parameter.  WDYT?
```
#gcal-persistence:calendar_name=
#gcal-persistence:offset=
#gcal-persistence:executescript=
```

watou commented:
> I didn't do the config.xml change, as I'd prefer to get this part right first. :-)

Fair enough.  Allowing GCal to be configured from Paper UI could be left to another PR if you prefer.  So I only would like to see the .cfg files made more "idiot-proof" for this PR.  Thanks, Mark!

mhilbush commented:
Latest commit has the following changes:
- updated gcal.cfg comments
- added gcal-persistence.cfg
- openhab_default.cfg now matches gcal.cfg and gcal-persistence.cfg

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -236,18 +236,33 @@ chart:provider=default
+# The filter criteria by which calendar events are searched. The Google Calendar 
+# API will do a text search to find calendar events that match the supplied terms. 
+# All calendar event fields are searched, except for extended properties (optional)
+#filter=
```
Prepend `gcal:`?

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -236,18 +236,33 @@ chart:provider=default
+# Refresh interval (in milliseconds) is the frequency with which the
+# Google calendar will be checked for calendar events (optional, defaults 
+# to 900000 [15 minutes])
+#refresh=
```
Prepend `gcal:`?

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -387,22 +402,22 @@ logging:pattern=%date{ISO8601} - %-25logger: %msg%n
-
-# the url of the calendar feed
-#gcal-persistence:url=
+# The following parameters are used to configure the GCal Persistence binding.
```
"service", not "binding"

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -387,22 +402,22 @@ logging:pattern=%date{ISO8601} - %-25logger: %msg%n
-# represents the State (optional, defaults to
-# '> if (PresenceSimulation.state == ON) sendCommand(%s,%s)')
+# represents the State (optional, defaults to 'send %s %s')
 #gcal-persistence:executescript=
```
This feature doesn't exist any more, does it?

mhilbush replied:
I left it in because it's still pulled from the config on line 116, and is used on line 166, in GCalPersistenceService.java

watou commented on features/openhab-addons-external/src/main/resources/conf/gcal.cfg:
```diff
@@ -1,10 +1,30 @@
-# Copied from Credentials page on console.developers.google.com
+################################### GCal Binding ###################################
+#
+# Before using the GCal bindings in openHAB2, you need to have a Google API
```
Not "bindings," more like "services."  Yes, it's a bit confusing. ;-)

mhilbush commented:
I've corrected everything except for `gcal-persistence:executescript=`.  Since it was still used in the code, I left it in the config.  WDYT?

Sorry for all the revisions...

watou commented:
Sorry for my mistake, @mhilbush.  The name `executescript` is indeed still used, but it's not actually capable of being a script fragment any more so I got myself confused by the old name.  Its only valid formats are now `send %s %s` or `update %s %s`.

mhilbush commented:
No worries.  Then I think I've incorporated all your feedback so far.

However...  I just tried to install gcal-persistence in OH2 and I got the below error.  Looks like gcal-persistence imports a bunch of ch.qos.logback.* packages.  Not familiar with that, but I can look into it.
```
2016-12-15 08:24:23.662 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/opt/openhab/addons/org.openhab.persistence.gcal-1.9.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.persistence.gcal [198]
  Unresolved requirement: Import-Package: ch.qos.logback.classic
        at org.eclipse.osgi.container.Module.start(Module.java:434)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:393)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1245)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1217)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:509)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:358)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:310)[8:org.apache.felix.fileinstall:3.5.0]
```

watou commented on distribution/openhabhome/configurations/openhab_default.cfg:
```diff
@@ -234,19 +234,34 @@ chart:provider=default
-#gcal:url=
+# This is the name you gave to your Google Calendar.  GCal will download calendar 
+# events from this calendar (required)
+#gcal:calendar_name=
```
The code says:
```
            String urlString = (String) config.get("calendar_name");
            if (!StringUtils.isBlank(urlString)) {
                calendar_name = urlString;
            } else {
                logger.error(
                        "gcal:calendar_name must be configured in openhab.cfg. Calendar name or word \"primary\"  MUST be specified");
                throw new ConfigurationException("calendar_name",
                        "gcal:calendar_name must be configured in openhab.cfg. Calendar name or word \"primary\"  MUST be specified");
            }
```
but the member `calendar_name` defaults to `"primary"` as well.  So the code might allow this to be optional.  For this PR, could you add that the special keyword `primary`" can be used?  Ideally it would be verified that specifying `calendar_name` is optional and defaults to `primary` and the code changed to allow it to be optional (and tested!). 😄

mhilbush replied:
I've confirmed the following:
- both `gcal `and `gcal-persistence` require a calendar name in the cfg.  If not supplied, it doesn't default to anything.  It does log errors in the log file, however.
- using the keyword `primary `does use your default Google calendar.  My default calendar is now full of presence simulation events.  :-(

Therefore, I'm not sure it's a good idea to default to `primary `if no calendar_name is specified (at least for `gcal-presence`), unless one wants to discover a bunch of presence simulation events in their default Google calendar.

I would propose that I update the cfg comments (gcal, gcal-persistence, and openhab_default) to reflect the usage of the keyword `primary`, but leave `calendar_name` required.

Also, in order to get gcal-persistence to run in OH2, I needed to remove these imported packages from `MANIFEST.MF`.  Should I include an updated `MANIFEST.MF` in this PR?
```
ch.qos.logback.classic,
ch.qos.logback.classic.encoder,
ch.qos.logback.classic.spi,
ch.qos.logback.core,
ch.qos.logback.core.encoder,
ch.qos.logback.core.pattern,
ch.qos.logback.core.spi,
```

watou replied:
> both gcal and gcal-persistence require a calendar name in the cfg. If not supplied, it doesn't default to anything. It does log errors in the log file, however.

The code does log the error and throw the exception, but the `calendar_name` is initialized to `"primary"` in the class so it didn't make sense -- either have no default for the member, or allow it to be optional.  In any case, I'm fine with requiring the user to explicitly set the calendar, for the reasons you gave.  Yes, please update the .cfg files to make it clear, and thanks!

> Should I include an updated MANIFEST.MF in this PR?

Beat you to it!  #4899 😄

watou commented:
Thanks for trying the persistence bundle on OH2, @mhilbush.  I've submitted https://github.com/openhab/openhab/pull/4899 to fix the problem you saw.  Consider replacing the bundle with [this one](https://dl.dropboxusercontent.com/u/4286376/gcal-persistence-logging/org.openhab.persistence.gcal-1.9.0-SNAPSHOT.jar) if you have a chance to test again.

mhilbush commented:
Confirmed that version works.  Thanks, @watou.

Are you ok with the above approach for updating the cfg file comments wrt primary keyword?

mhilbush commented:
Just pushed the changes re: primary keyword.  Anything else you thing needs to be done on this?

watou commented:
LGTM!  Anything else from your side before I merge?

mhilbush commented:
Nope.  I'm good.

mhilbush commented:
BTW, thanks for your help and feedback.

watou commented:
Thanks for your contribution!  It's great to see GCal up and usable again with the group effort! 😄

mhilbush commented:
It looks like there are some minor edits needed to the GCal Wiki page.  Any issue with me making those changes?

watou commented:
Please go right ahead!  Having all questions answered by clear documentation is really the most important thing of all!

watou commented:
Big thanks again to @SirAd for making it possible with the OAuth2 update.

mhilbush commented:
@watou  I must've done something wrong in feature.xml.  gcal-persistence is showing up in PaperUI, but not gcal.  This is the feature name I gave it.

`
<feature name="openhab-io-gcal1" description="GCal" version="${project.version}">
`

Should it be `openhab-binding-gcal1`, and should it be up in the binding section?

watou commented:
Is the service not showing up under the Misc. Tab?

mhilbush commented:
Correct.  Not showing under any tab (including misc).

watou commented:
In several hours I will be able to have a look.

mhilbush commented:
From what I can see, currently the only features showing in Misc are from OH2.  

Do we need a misc section in the openhab feature.xml, then change the gcal entry to:
```
<feature name="openhab-misc-gcal1" description="GCal" version="${project.version}">
    <feature>openhab-runtime-base</feature>
    <feature>openhab-runtime-compat1x</feature>
    <bundle start-level="80">mvn:org.openhab.io/org.openhab.io.gcal/${project.version}</bundle>
    <configfile finalname="${openhab.conf}/services/gcal.cfg" override="false">mvn:${project.groupId}/openhab-addons-external/${project.version}/cfg/gcal</configfile>
</feature>
```

And change gcal-persistence to:
```
<feature name="openhab-persistence-gcal" description="GCal Persistence" version="${project.version}">
    <feature>openhab-runtime-base</feature>
    <feature>openhab-runtime-compat1x</feature>
    <feature>openhab-misc-gcal1</feature>
    <bundle start-level="80">mvn:org.openhab.persistence/org.openhab.persistence.gcal/${project.version}</bundle>
    <configfile finalname="${openhab.conf}/services/gcal-persistence.cfg" override="false">mvn:${project.groupId}/openhab-addons-external/${project.version}/cfg/gcal-persistence</configfile>
</feature>
```

watou commented:
My understanding is that it's done by convention, so your change makes sense.  I will delve deeper to find out for sure.

kaikreuzer commented:
`openhab-misc-gcal1` will indeed make it appear on the "MISC" tab under Extensions in the Paper UI - this should be used for any feature that is installable by the user, while `openhab-io...` are only shared bundles that are required by other features, but do not make sense on their own.

But if you expose it to the UI, you should think of a slighly longer name than "GCal" for it, please!

watou commented:
> But if you expose it to the UI, you should think of a slighly longer name than "GCal" for it, please!

I was just about to suggest that its name be changed to a name that makes it clear 1) what is being integrated into openHAB, and 2) what function it performs.  Something like "Google Calendar Scheduler" or "Google Calendar Connector".

mhilbush commented:
I like Google Calendar Scheduler.

Since OH1 users already know it by GCal, could it be this, as there appears to be precedent for having a parenthetical in the name:
`Google Calendar (GCal) Scheduler`

And, for gcal-persistence, it could be one of these, which is more descriptive of what it actually does:
```
Google Calendar Presence Simulation
Google Calendar (GCal) Presence Simulation
```

mhilbush commented:
Or, to be a bit more consistent with Scheduler...
```
Google Calendar Presence Simulator
Google Calendar (GCal) Presence Simulator
```

watou commented:
I don't see a need for (GCal) but I like your suggestions.

mhilbush commented:
Alright, if there are no objections, I'll do:
```
Google Calendar Scheduler
Google Calendar Presence Simulator
```
